### ISSUE TITLE
Multi-language support in RVD

### ISSUE THREAD

**Issue by otsakir** [Mobicents/RestComm#190]
Add multi-language support in RVD UI. An angular based solution with a separate file with all translation strings and symbolic constants that refer to these.

This angular module could be used:

http://www.ng-newsletter.com/posts/angular-translate.html

**Comment by ivelin** [comment]
@otsakir , we have a new contributor who is interested in taking on this task.
Please meet @ghjansen.

@ghjansen, feel free to reference @otsakir, @gvagenas or @brainslog when you ask questions related to this issue.

**Comment by otsakir** [comment]
That's great @ghjansen, any contribution will be most valuable!

I suggest branching out of rvd_master branch which has the latest RVD additions.

**Comment by ghjansen** [comment]
Thank you @ivelin and @otsakir ! Its nice to be here with you guys!

Following @otsakir suggestion, I will push the new branch and start the issue this weekend.

**Comment by otsakir** [comment]
Ok @ghjansen, reference me if you need any help.

**Comment by ghjansen** [comment]
@otsakir could you check my commit on Issue190_MultiLanguageSupport to ensure that everything is going as expected? I've done at index and login.

Using the source bellow at index page, you can switch between the languages in runtime:

<div ng-controller="translateController">
	<button ng-click="changeLanguage('ptBR')">ptBR</button>
	<button ng-click="changeLanguage('enUS')">enUS</button>
</div>

I'm also translating it to my local language, hope there is not a problem for doing that.

I wait your confirmation to go ahead with the next pages.

**Comment by otsakir** [comment]
Great job @ghjansen. Just tested the patch. It works perfectly and is clean. 

No problem at all with your local language. Adding other language will be a breeze.

Feel free to proceed and don't hesitate to ask if in trouble.

Thanks.

**Comment by ivelin** [comment]
nice progress @ghjansen

**Comment by ghjansen** [comment]
Thank you @otsakir and @ivelin !
I will send the other parts as soon as I finish.

**Comment by ghjansen** [comment]
Hi @ivelin and @otsakir , both internationalization and translation were done.
A few terms were omitted since they are internally processed. I can enumerate them if necessary.
The translation was made considering the Portuguese Language Orthographic Agreement of 1990, in use since 2009.
No unit-tests were implemented yet. Can you give me directions to the appropriate resource/tool I need to use to make them?

Please, let me know if there is something I need to change about the last commits or anything else.

**Comment by deruelle** [comment]
Congrats @ghjansen ! 
Do you think you can also work with @croufay on creating a tutorial for easily adding more languages to RestComm for people ready to contribute internationalization ?

**Comment by otsakir** [comment]
That's good news @ghjansen ! I will check the commits and get back to you. 

On unit-testing, that's ok. RVD is a relatively young project with very active development until recently and a lot of refactoring. With that in mind, unit-testing was spared until a time when things would settle. That time has arrived.

If you can provide tests running on the well known angular testing frameworks like karma, ng-Mock or jasmine it would be just fine. Otherwise, i can let you know when RVD's got its test suite and resume adding tests then.

**Comment by ghjansen** [comment]
Thank you @deruelle , the tutorial is a good idea! It will be nice to help on it also. Just give me the directions to start.
Thank for your instructions @otsakir . I researched about these frameworks on the weekend, now its time to go a little deeper. I will let you know if I need some help, but i'm thinking that will be no problem.
I'm a little busy at the university until next week, some of my replies may delay. Sorry in advance about that.

**Comment by otsakir** [comment]
Sure @ghjansen . I'll be glad to help. Take your time and engage when you see fit.

**Comment by deruelle** [comment]
@croufay, @desipepova can you create a guest account for @ghjansen http://docs.telestax.com/restcomm-pages/ so he can contribute a tutorial for your review on internationalizing RestComm ?

**Comment by otsakir** [comment]
btw, @ghjansen, could you enumerate the terms that you found that are internally processed?

**Comment by desipepova** [comment]
@ghjansen your account in http://docs.telestax.com has been created. You should have received an email with account details.

**Comment by ghjansen** [comment]
Thank you @desipepova , I logged successfully.
@otsakir I will collect some samples about this case and bring them to you as soon as I can. Probably at the weekend.

**Comment by ivelin** [comment]
Congrats @ghjansen ! Once this issue is closed we can proceed with more challenging tasks.

**Comment by ghjansen** [comment]
Thats cool @ivelin ! Thank you! I can guarantee I'm excited to go ahead!

@desipepova , I noticed that after login to create the tutorial there is no 'Posts' tab in the left menu, as shown by the image bellow. Could you confirm if some missing permission could be causing that? Please, let me know if there is another way to create the tutorial without the 'Posts' tab.

@otsakir , I know I should take a deeper look before share with you, but considering time improvements for us both, here is two examples of what i mentioned as "internally processed" terms:

1 - The step name placed between the step-heading tags on each *Step page. I know the labels come from steps.js but dont know yet how to translate them;
2 - Option values in some select elements (the ones using ng-options), like Yes/No in recordStep or object/array/value in externalServiceStep.

Translate such terms may not be a problem, but I need to find out the proper way to do it.

I'm also writing some adjustments that I've noticed needed in the portuguese translation, as in other parts. Some things I know that could be better.

Thank you all!
![screen shot 2015-03-19 at 20 50 42](https://cloud.githubusercontent.com/assets/1459428/6744753/dc1d2410-ce8b-11e4-8ff6-aa73d36a7c59.png)

**Comment by otsakir** [comment]
Ok @ghjansen , i see what you mean. 

Are you done with the rest of the translation issue? If yes, I can proceed to review and merge.

**Comment by ghjansen** [comment]
Not yet @otsakir , I can only work at the weekends. Tomorrow I will start the tests and the tutorial. I inform you when finish.

**Comment by ghjansen** [comment]
Hi @otsakir , I'm trying to make some tests using Karma Jasmine but there is some issues caused by the design of angular-translate that requires an adaptation of our current test suite. The main change is to overwrite the module configuration at runtime. I'm afraid this could directly affect the other tests and take more time than expected to adapt.

For more info, please see: https://github.com/angular-translate/angular-translate/issues/42 and http://angular-translate.github.io/docs/#/guide/22_unit-testing-with-angular-translate

Please, let me know if i can to go ahead this way, or maybe temporarily avoid the translation tests.

**Comment by ghjansen** [comment]
@otsakir i've done the last push with adjustments, including the RFC 5646 pattern to the language file names.
@deruelle the tutorial is ready. I'm just waiting the for the confirmation of @desipepova about the Posts tab, as I described some comments before, then I will post it.
Please, let me know if there is anything more I can do to this issue.

**Comment by otsakir** [comment]
Thanks @ghjansen . I will proceed with review and merging.

No need to proceed with unit-test implementation if there are issues. There is probably a bigger plan that needs to be caried out before RVD easily supports pluggable testing in the client side.

**Comment by otsakir** [comment]
@ghjansen , your patch is merged into rvd_master and everything works ok. There have been some minor changes in translation strings that have changed since  you started working on it. For the portugeese versions of the strings i've added the 'TRANSLATE' prefix so that they stand out are easily spotted. 

There is also a new issue about "remembering" the language that seems appropriate in case you feel like working further on multilingual RVD capability. If you do, just make sure you have pulled all changes from rvd_master branch before starting to make modifications. It will save re-merging old changes again.

https://github.com/Mobicents/RestComm/issues/199

**Comment by ghjansen** [comment]
Thank you @otsakir ! I've pulled all changes from rvd_master and already checked the prefixes in the portuguese language file.
It will be nice to go further with #199, sure I accept it!
Can I make both translation and #199 at rvd_master or should I start in a new branch from it?

**Comment by otsakir** [comment]
That's great @ghjansen! Go ahead and reference me if you need any help.

It's better to start working in a new issue. There is a small time overhead to set it up and sync with rvd_master but in case things end up messy, it can save a lot of time.

**Comment by ghjansen** [comment]
Ok @otsakir , thank you! I will do it as soon as I can!

**Comment by ivelin** [comment]
@otsakir should this issue be closed now?
